### ISSUE TITLE
Unity can Still Stall

### ISSUE THREAD

**Issue by DoubleStyx** [Frozenreflex/Thundagun#19]
There are some operations, like world loading, that cause Unity to stall. I'm not sure exactly why this is, but it might be worth looking into. I assume it's related to the number of updates that Unity has to process on world load? If so, then there's not much we can really do about this other than spreading out the changes over time.

Which is something I actually did not think about until now.

With sync mode, we guarantee Unity and Resonite each run their cycles in full before they can begin another cycle.
With async, we at least guarantee that Unity only renders full states.
But perhaps there could be a third option, which doesn't guarantee batching or locking. We could potentially use this for very heavy workloads, like loading worlds.

At this point, these mod options are getting complicated enough that I'm not sure we really need the distinction anymore.

Basically, it would be a triage system:

Sync is the default option as it does not use half-states or state skipping/reusing.
Async is a backup that uses full states for stability/determinism but can skip/reuse frames to smoothen things.
Desync is a secondary backup that could handle both extreme Resonite stalls (100:1, where you might still want to see the game advance in some way) and extreme Unity stalls (1:100, where you need frames to survive but are okay with changes getting processed over time instead of all at once).

So we set a sync-to-async threshold and an async-to-desync threshold.

**Comment by DoubleStyx** [comment]
I'd say that with the introduction of the desync behavior/threshold, we'd remove the auto mode and sync mode options entirely. Instead, the mod would just expose two thresholds and the EMA exponent. In practice, you only really want to use sync mode; async just kinda looks bad, though not as bad as desync for realtime use. And desync would be far worse. So having all of this behavior be dynamic and configurable seems like the logical approach here.

**Comment by DoubleStyx** [comment]
In the case of extreme Resonite stalls (1:64), we need to limit the amount of work Resonite is doing, most likely by using Unity's EMA; we can't really stall Resonite here, but we can continue on without it (unless we want to actually block the thread). But for extreme Unity stalls (1:64), we need to limit the amount of work Unity is doing, most likely by using Resonite's frame EMA.

The easiest approach would probably be to work for `time = ResoniteEMA`. This is sort of a starting point. Then on the next Unity update, we calculate `offset = (DateTime.Now() - UnityStartTime) - time` to get the amount of time that went over budget. Then we set time = `time - offset`, which generates a new budget that fits within the device's refresh rate.

I still don't know if I like this, though. I suppose it's less about hitting the device refresh rate and more about parity between Resonite and Unity's timing. 

Sync mode forces the two to be the same. Async simulates this by using skipped/stale states. Desync should let Unity run for as long as Resonite following the same principle.

What is Resonite actually doing during a Unity stall? Is it stalling itself? 

I might try this approach though and see how well it works.

I guess the distinction here is whether the ratio is more important or the framerate is more important.

**Comment by DoubleStyx** [comment]
If we want to block on the thread level, then we'd have to wait the competing thread for as long as the faster thread's EMA. Though I don't know if we can really block the Unity thread since we don't own it?

To simulate thread ownership on the Unity side, we'll just use the dynamic offset calculation. For the Resonite side, it sort of doesn't matter at that point, but for consistency we could technically sleep that thread for a bit.

**Comment by DoubleStyx** [comment]
The current approach is kind of weird, though. It sets the mode purely based on ratios, not absolute values. We could also expose a maxInterval setting for the worst case, where both are stalling and we need them to continue.

It's kind of pointless though. We can't sleep the Resonite thread. So we can only really "sleep" the Unity thread by giving it less work to do by departing early. 

Right now, if Resonite is 64:1, we can't sleep it, but we can just process whatever's there since there will be 1/64th of the workload. If Unity is 64:1 from change processing, we can't sleep it directly either, but we can use the dynamic departure to give Unity 1/64th of the expected workload.

If Unity is 64:1 from rendering itself, then glhf I guess.

**Comment by DoubleStyx** [comment]
Closing as this is basically encompassed by the pr.
